classes/webservice/WebserviceSpecificManagementAttachments.php

```diff
@@ -0,0 +1,316 @@
+            }
+        }
+        // Need to set an object for the WebserviceOutputBuilder object in any case
+        // because schema need to get webserviceParameters of this object
+        if (isset($object)) {
+            $this->wsObject->objects['empty'] = $object;
+        }
+    }
+
+    /**
+     * Handles attachment file download
+     *
+     * @throws WebserviceException if attachment is not existing or file not available
+     *
+     * @return string[] file details
+     */
+    public function executeFileGetAndHead()
+    {
+        $attachment = new Attachment((int) $this->wsObject->urlSegment[2]);
+        if (!$attachment) {
+            throw new WebserviceException(
+                sprintf(
+                    'Attachment %d not found',
+                    $this->wsObject->urlSegment[2]
+                ),
+                [
+                    1,
+                    Response::HTTP_INTERNAL_SERVER_ERROR,
+                ]
+            );
+        }
+
+        // Physical file location
+        $file = _PS_DOWNLOAD_DIR_ . $attachment->file;
+        // Check if file exists
+        if (!file_exists($file)) {
+            throw new WebserviceException(
+                sprintf(
+                    'Unable to load the attachment file for attachment %d',
+                    $this->wsObject->urlSegment[2]
+                ),
+                [
+                    1,
+                    Response::HTTP_INTERNAL_SERVER_ERROR,
+                ]
+            );
+        }
+
+        // Return file details
+        return array(
+            'file' => $file,
+            'mime' => $attachment->mime,
+            'file_name' => $attachment->file_name,
+            'file_size' => $attachment->file_size,
+        );
+    }
+
+    /**
+     * Handles file upload
+     *
+     * Creates new attachment or replaces existing with a new file.
+     * [PUT] update existing attachment file
+     * [POST] create new attachment
+     */
+    public function executeFileAddAndEdit()
+    {
+        // Load attachment with or without id depending on method
+        $attachment = new Attachment($this->wsObject->method == 'PUT' ? (int) $this->wsObject->urlSegment[1] : null);
+
+        // Check form data
+        if (isset($_FILES['file']) && is_uploaded_file($_FILES['file']['tmp_name'])) {
+            // Ensure file is within allowed size limit
+            if ($_FILES['file']['size'] > (Configuration::get('PS_ATTACHMENT_MAXIMUM_SIZE') * 1024 * 1024)) {
+                $this->wsObject->errors[] = sprintf(
+                    $this->l('The file is too large. Maximum size allowed is: %1$d kB. The file you are trying to upload is %2$d kB.'),
+                    Configuration::get('PS_ATTACHMENT_MAXIMUM_SIZE') * 1024,
+                    number_format(($_FILES['file']['size'] / 1024), 2, '.', '')
+                );
+            } else {
+                // Assign unique id
+                $uniqid = sha1(uniqid()); // must be a sha1
```

Comment:
I think you misunderstood me here, should be something like.

if($attachment->id) {
    $uniqid = $attachment->file;
} else {
    do {
        $uniqid = sha1(uniqid()); // must be a sha1
    } while (file_exists(_PS_DOWNLOAD_DIR_ . $uniqid));
}

Reply:
Yes, and it genererated a new id until it found one that did not match a previous file.
Without the loop, it MAY (if you are tremendously unlucky) generate a string that was previously used for another attachment and overwrite that file.

Reply:
Done!

Reply:
What about an md5 checksum of the attachment file?

Reply:
Looks good

Reply:
We are not looking for validate the checksum of the file, just creating a random file name.

Reply:
Yes, but unique right? the md5 checksum would give you a unique file name.

Reply:
md5 checksum are not uniq (can be hacked to replace an other file) and we will not be able to send the same attachment twice. More it consume more memory to read the file to create the checksum than playing with uniqid.